crates/adapters/src/controller.rs

```diff
@@ -807,6 +807,144 @@ impl Controller {
+            tokio::select! {
+                _ = child.wait() => {}
+                _ = tokio::time::sleep(Duration::from_secs(duration)) => {
+                    // Send SIGINT to the samply process to stop recording.
+                    nix::sys::signal::kill(
+                        nix::unistd::Pid::from_raw(child_pid as i32),
+                        nix::sys::signal::Signal::SIGINT,
+                    )
+                    .context("failed to send SIGINT to samply process")?;
+                }
+            }
+
+            let output = child
+                .wait_with_output()
+                .await
+                .context("failed when waiting for samply process")?;
+
+            if !output.status.success() {
+                anyhow::bail!(
+                    "samply process failed with status: `{}`, samply stdout: `{}`, samply stderr: `{}`",
+                    output.status,
+                    String::from_utf8_lossy(&output.stdout).trim().replace("\n", "\\n"),
+                    String::from_utf8_lossy(&output.stderr).trim().replace("\n", "\\n"),
+                );
+            }
+
+            let mut profile_buf = Vec::new();
+
+            BufReader::new(File::open(profile_file).await.context(format!(
+                "failed to open samply profile file `{profile_file}`"
+            ))?)
+            .read_to_end(&mut profile_buf)
+            .await
+            .context(format!(
+                "failed to read samply profile file `{profile_file}`"
+            ))?;
+
+            let mut zip_buf = Vec::with_capacity(profile_buf.len() * 2);
+            let mut zip = zip::ZipWriter::new(std::io::Cursor::new(&mut zip_buf));
+            let options = zip::write::FileOptions::default()
+                .compression_method(zip::CompressionMethod::Deflated);
+            zip.start_file("profile.json", options)
+                .context("failed to start writing zip file for samply profile")?;
+            zip.write_all(&profile_buf)
+                .context("failed to write samply profile data zip")?;
+
+            zip.start_file("profile.syms.json", options)
+                .context("failed to start writing zip file for samply symbols")?;
+
+            let mut syms_buf = Vec::new();
+            let symbols_file = format!("{}.syms.json", profile_file);
+
+            let mut syms_reader = BufReader::new(File::open(symbols_file).await.context(
+                format!("failed to open samply symbols file `{profile_file}`"),
+            )?);
+            syms_reader
+                .read_to_end(&mut syms_buf)
+                .await
+                .context(format!(
+                    "failed to read samply symbols file `{profile_file}`"
+                ))?;
+            zip.write_all(&syms_buf)
+                .context("failed to write samply symbols data zip")?;
+
+            tracing::debug!("collected symbols for samply: {} bytes", syms_buf.len());
+
+            drop(zip);
```

Comment:
this seems unnecessary?

Reply:
It is necessary to return the file buf later.